File: rust/examples/vmcall.rs
//! Example demonstrating vmcall - calling guest functions from host

use libloong::{Machine, MachineOptions};
use std::env;
use std::fs;
use std::process;

fn main() {
    // Get ELF path from command line
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!(
            "Usage: {} <loongarch_elf_file> [function_name] [args...]",
            args[0]
        );
        eprintln!();
        eprintln!("Example:");
        eprintln!("  {} ../tests/programs/cxx_test.elf test_addition", args[0]);
        process::exit(1);
    }

    let elf_path = &args[1];
    let func_name = args.get(2).map(|s| s.as_str()).unwrap_or("main");

    // Parse remaining arguments as u64
    let func_args: Vec<u64> = args[3..].iter().filter_map(|s| s.parse().ok()).collect();

    // Read the ELF binary
    let binary = match fs::read(elf_path) {
        Ok(data) => data,
        Err(e) => {
            eprintln!("Failed to read ELF file '{}': {}", elf_path, e);
            process::exit(1);
        }
    };

    println!("Loading LoongArch ELF: {}", elf_path);

    // Create machine
    let mut machine = match Machine::new(&binary, MachineOptions::default()) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("Failed to create machine: {}", e);
            process::exit(1);
        }
    };

    // Setup Linux environment
    Machine::setup_linux_syscalls();
    if let Err(e) = machine.setup_linux(&["program"], &[]) {
        eprintln!("Failed to setup Linux: {}", e);
        process::exit(1);
    }

    // Setup stdout callback
    libloong::set_stdout_callback(Some(|data| {
        print!("{}", String::from_utf8_lossy(data));
    }));

    // Check if function exists
    if !machine.has_symbol(func_name) {
        eprintln!("Function '{}' not found in binary", func_name);
        eprintln!("\nTry running with a different function name.");
        process::exit(1);
    }

    let func_addr = machine.address_of(func_name);
    println!(
        "Calling function '{}' at address 0x{:x}",
        func_name, func_addr
    );
    println!("Arguments: {:?}", func_args);
    println!();

    // Call the guest function
    match machine.vmcall_by_name(func_name, &func_args) {
        Ok(result) => {
            println!("\nFunction returned: {} (0x{:x})", result, result);
            println!("Instructions executed: {}", machine.instruction_counter());
        }
        Err(e) => {
            eprintln!("\nVmcall failed: {}", e);
            eprintln!("Instructions executed: {}", machine.instruction_counter());
            eprintln!("PC: 0x{:016x}", machine.get_pc());
            process::exit(1);
        }
    }
}

File: go/examples/vmcall.go
// Example demonstrating vmcall - calling guest functions from host
package main

import (
	"fmt"
	"os"
	"strconv"

	libloong "github.com/gonz/libloong/go"
)

func main() {
	// Get ELF path from command line
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <loongarch_elf_file> [function_name] [args...]\n", os.Args[0])
		fmt.Fprintln(os.Stderr)
		fmt.Fprintln(os.Stderr, "Example:")
		fmt.Fprintf(os.Stderr, "  %s ../tests/programs/cxx_test.elf test_addition\n", os.Args[0])
		os.Exit(1)
	}

	elfPath := os.Args[1]
	funcName := "test_function" // Don't use "main" or "_start" - only regular functions after init
	if len(os.Args) > 2 {
		funcName = os.Args[2]
	}

	// Parse remaining arguments as uint64
	var funcArgs []uint64
	for _, arg := range os.Args[3:] {
		if val, err := strconv.ParseUint(arg, 0, 64); err == nil {
			funcArgs = append(funcArgs, val)
		}
	}

	// Read the ELF binary
	binary, err := os.ReadFile(elfPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read ELF file '%s': %v\n", elfPath, err)
		os.Exit(1)
	}

	fmt.Printf("Loading LoongArch ELF: %s\n", elfPath)

	// Create machine
	machine, err := libloong.NewMachine(binary, nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create machine: %v\n", err)
		os.Exit(1)
	}
	defer machine.Close()

	// Setup Linux environment
	libloong.SetupLinuxSyscalls()
	if err := machine.SetupLinux([]string{"program"}, []string{}); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to setup Linux: %v\n", err)
		os.Exit(1)
	}

	// Check if function exists
	if !machine.HasSymbol(funcName) {
		fmt.Fprintf(os.Stderr, "Function '%s' not found in binary\n", funcName)
		fmt.Fprintln(os.Stderr, "\nNOTE: You can only vmcall regular functions, not 'main' or '_start'.")
		fmt.Fprintln(os.Stderr, "      These entry points need full initialization.")
		fmt.Fprintln(os.Stderr, "      Use Simulate() to run the full program instead.")
		os.Exit(1)
	}

	fmt.Printf("Calling function '%s'\n", funcName)
	fmt.Printf("Arguments: %v\n", funcArgs)
	fmt.Println()
	fmt.Println("NOTE: This calls the function AFTER the program has initialized.")

	// Call the guest function
	result, err := machine.VMCallByName(funcName, ^uint64(0), funcArgs...)
	if err != nil {
		fmt.Fprintf(os.Stderr, "\nVmcall failed: %v\n", err)
		fmt.Fprintf(os.Stderr, "Instructions executed: %d\n", machine.InstructionCounter())
		fmt.Fprintf(os.Stderr, "PC: 0x%016x\n", machine.GetPC())
		os.Exit(1)
	}

	fmt.Printf("\nFunction returned: %d (0x%x)\n", result, result)
	fmt.Printf("Instructions executed: %d\n", machine.InstructionCounter())
}
